### PR TITLE
Dynamic circuit break period

### PR DESCRIPTION
**Pull request by Brondahl**
Initial work to allow for dynamically calculated CircuitBreaker break periods.

### The issue or feature being addressed
See Issue: https://github.com/App-vNext/Polly/issues/653
( @reisenberger )

### Details on the issue fix or feature implementation
See Issue: https://github.com/App-vNext/Polly/issues/653
Note that there seem to be 2 contentious design choices outstanding; mentioned on that Issue.

w.r.t. Non-Positive inputs, I've gone with leaving static-duration behaviour unchanged, and having no checks for dynamic durations.

w.r.t. Delay durations when the response is slow to come back, I've gone with re-using the same delay, until one of the responses actually completes.

Several things still need to be done, before this is fully suitable for deployment, but they're all fairly rote changes, so I want to get the meat of the change approved before doing them, since if there are substantial markups the rote changes might need tweaks, which would be ... tedious :)

* Trivially apply all the same changes to the Factory Methods for other syntaxes (CircuitBreakerAsync, CBTReslt, CBTResultAsync)
* Figure out what (if anything?) should be done for the Advanced version of the CB. Then apply all the same boring Factory Method changes to all 4 version of that.
* Update existing tests for all of the above.
* Create new tests for the specific functionality for the duration to change dynamically (as opposed to the existing tests which all use constant-return functions. This ensures that the existing functionality is unchanged, and that the most basic use-case of the functionality works, but is somewhat lacking in depth.)
* Add general documentation (the XMLDocs are updated, but I assume there's a high-level doc somewhere that I should update too.

### Confirm the following

**FIXED**: ~Branch state on your repo currently looks weird~

^ See above, the existing tests have been updated, and crudely exercise the new functionality. Additional tests are required, and I'll create them in due course; wanted to get the meat of the change up for someone to look at while I'm doing that.

- [x]  I started this PR by branching from the head of the latest dev vX.Y branch, or I have rebased on the latest dev vX.Y branch, or I have merged the latest changes from the dev vX.Y branch
- [x]  I have targeted the PR to merge into the latest dev vX.Y branch as the base branch
- [Y^]  I have included unit tests for the issue/feature
- [Y]  I have successfully run a local build

### CODE REVIEW

**Comment by Brondahl** [comment]
Note that you may well find it easier to review this commit-by-commit. The first commit is entirely mechanical copy-pasting; the ones after that are more interesting.

**Comment by reisenberger** [comment]
Thanks @Brondahl for the contribution!  I am now on family/vacation time; I might get to look at this in the next couple days, but it could also be late December or January.  Thanks.

**Comment by Brondahl** [comment]
👍 
Enjoy your break.
Thanks for the expectation-setting :)

File: src/Polly/CircuitBreaker/ConsecutiveCountCircuitController.cs [hunk 1]
```diff
@@ -23,7 +24,8 @@ public override void OnCircuitReset(Context context)
         {
             using (TimedLock.Lock(_lock))
             {
-                _consecutiveFailureCount = 0;
+                _consecutiveFailures = 0;
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/ConsecutiveCountCircuitController.cs]: Good point. Ended up on:
```
        private int _consecutiveFailuresFromClosedState;
        private int _consecutiveFailuresFromHalfOpenState;
```

File: src/Polly/CircuitBreaker/CircuitStateController.cs [hunk 2]
```diff
@@ -126,14 +133,32 @@ protected void ResetInternal_NeedsLock(Context context)
...
-                // It's time to permit a / another trial call in the half-open state ...
-                // ... but to prevent race conditions/multiple calls, we have to ensure only _one_ thread wins the race to own this next call.
-                return Interlocked.CompareExchange(ref _blockedTill, SystemClock.UtcNow().Ticks + _durationOfBreak.Ticks, currentlyBlockedUntil) == currentlyBlockedUntil;
+                return false;
             }
-            return false;
+
+            // It's time to permit a / another trial call in the half-open state ...
+            // ... but to prevent race conditions/multiple calls, we have to ensure only _one_ thread wins the race to own this next call.
+            // If "we" are the thread that wins that races, then we also want to update the BlockedTill date, so that no other thread / call tries to execute before this Test has reached *it's* conclusion.
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: done

File: src/Polly/CircuitBreaker/CircuitStateController.cs [hunk 3]
```diff
@@ -94,10 +95,16 @@ public void Isolate()
             }
         }
 
-        protected void Break_NeedsLock(Context context) => BreakFor_NeedsLock(_durationOfBreak, context);
+        protected void Break_NeedsLock(int consecutiveHalfOpenFailures, Context context)
+        {
+            var nextBreakDuration = _factoryForNextBreakDuration(consecutiveHalfOpenFailures);
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Cool. Works for me.

Do you actively prefer a different `Exception` type here (for the dynamically calculated duration), than for the static version which throws a `ArgumentOutOfRangeException`.

I can see arguments in either direction:
* It's *logically* the same flaw, so use the same Exception.
* It's is programatically definitely *NOT* an argument to a function, at the point that we'd be throwing it; so we shouldn't mis-use that Exception type.

File: src/Polly/CircuitBreaker/CircuitStateController.cs [hunk 4]
```diff
@@ -126,14 +133,32 @@ protected void ResetInternal_NeedsLock(Context context)
 
         protected bool PermitHalfOpenCircuitTest()
         {
-            long currentlyBlockedUntil = _blockedTill;
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: 👍 . Done. Also renamed everything from "Till" to "Until". :)

File: src/Polly/CircuitBreaker/CircuitStateController.cs [hunk 5]
```diff
@@ -126,14 +133,32 @@ protected void ResetInternal_NeedsLock(Context context)
...
+            // If "we" are the thread that wins that races, then we also want to update the BlockedTill date, so that no other thread / call tries to execute before this Test has reached *it's* conclusion.
+            //
+            // So combine doing those two things into an Interlocked.CompareExchange call.
+            // Attempt to update _blockedTill. Then examine the return value to determine whether someone else had beaten us to the punch.
+            // If not, then that tells us that we won the race, and we can permit the Circuit Test.
+            // If someone had already modified _blockedTil, then we lost the race and thus aren't going to be doing the Test.
+            //
+            // Note that if the Circuit Test is a failure, it will re-configure the Break limit, using the updated number of Test failures.
+            // Which means that it doesn't really matter what delay we put in here as long as no other attempts get made before this Test completes.
+            // This is helpful, since we don't have convenient access to the number of Test Failures here, and thus can't calculate the *correct* next delay.
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: I see what you're driving at.

I've applied a few further tweaks to your wording above. How's this:
```
            // Note that in general, we want to extend the blocked period by an amount based on the number of failures in halfOpen state (tracked elsewhere, last time a failure was received).  
            // The logic of halfOpen state is that it permits one fresh trial call per expiry of each extension-of-broken-state.
            // This particular extension of the Open state is occurring due to the preceding Open period expiring, or a halfOpen period expiring with no call result received.
            // i.e. it's not an extension triggered by any additional failures received.
            // Consequently, the period to extend by is unchanged from the last period, so we can reuse _currentBlockDuration (calculated last time we saw an active failure).
```

[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: > (Note: It does matter what delay we put in here, ...

Ah, yes. Agreed.
I initially thought it didn't matter at all and we could just put `Ticks.MaxValue` in. But you have UnitTests which insist that that isn't valid, so I changed it back.

But the comment was written the first time I tried to understand what was going on, and only got 80% of the understanding.

[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: I see what you're driving at.

I've applied a few further tweaks to your wording above. How's this:
```
            // Note that, in general, we want to extend the blocked period by an amount based on the number of failures in halfOpen state (tracked elsewhere, last time a failure was received).  
            // The logic of halfOpen state is that it permits one fresh trial call per expiry of each extension-of-broken-state.
            // This particular extension of the Open state is occurring due to the preceding Open period expiring, or a halfOpen period expiring with no call result received.
            // i.e. it's not an extension triggered by any additional failures received.
            // Consequently, the period to extend by is unchanged from the last period, so we can reuse _currentBlockDuration (calculated last time we saw an active failure).
```

File: src/Polly/CircuitBreaker/CircuitStateController.cs [hunk 6]
```diff
@@ -126,14 +133,32 @@ protected void ResetInternal_NeedsLock(Context context)
...
+            //
+            // Lastly note that the conceptual edge case of _blockedTil being equal to DateTime.MaxValue.Ticks, is not a concern, since in that case the
+            // if-check above will not have passed, so we won't have gotten here in the first place.
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Fair points.

So I've started by factoring out that "extend it, capping at the limit, if necessary":

```
        private void BreakFor_NeedsLock(TimeSpan durationOfBreak, Context context)
        {
            var (breakEnd, breakLength) = ExtensionUpToMaxValue(SystemClock.UtcNow(), durationOfBreak);

            _currentBlockDuration = breakLength;
            _blockedUntil = breakEnd.Ticks;

            var transitionedState = _circuitState;
            _circuitState = CircuitState.Open;

            _onBreak(_lastOutcome, transitionedState, durationOfBreak, context);
        }

        private Tuple<DateTime, TimeSpan> ExtensionUpToMaxValue(DateTime startPoint, TimeSpan proposedExtension, DateTime? maxDateTime = null)
        {
            var maxPossibleExtension = ((maxDateTime ?? DateTime.MaxValue) - startPoint);
            var extensionWouldTakeUsPastDateTimeMaxValue = proposedExtension > maxPossibleExtension;

            if (extensionWouldTakeUsPastDateTimeMaxValue)
            {
                return Tuple.Create(DateTime.MaxValue, maxPossibleExtension);
            }
            else
            {
                return Tuple.Create(startPoint + proposedExtension, proposedExtension);
            }
        }
```

[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Then add an overload to accept a Ticks-based StartDate. (We *could* do that calculation in PermitTest, but that method is already hella-complicated, so I'm inclined to avoid adding additional date-representation-faffing complexity to it.)

```
        private Tuple<DateTime, TimeSpan> ExtensionUpToMaxValue(long startPointTicks, TimeSpan proposedExtension, DateTime? maxDateTime = null)
        {
            var startDate = DateTime.MinValue + TimeSpan.FromTicks(startPointTicks);
            return ExtensionUpToMaxValue(startDate, proposedExtension, maxDateTime);
        }
```

[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Then use that in Permit...Test:

```
var (proposedNewBlockDeadline,_) = ExtensionUpToMaxValue(blockedUntil_AtStartOfPermitHalfOpenCircuitTest, _currentBlockDuration);
```

[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: All good?

**Comment by Brondahl** [comment]
Cool, so status update:
I've applied almost all the review markups, locally.

Tasks still outstanding:
* (@reisenberger) Sign-off on general review discussions.
* (@reisenberger) Confirm desired Exception type for `durationFactory() returns < 0" review markup.
* (@Brondahl) Implement ☝️ Exception throw.
* (@reisenberger + @Brondahl) Decide on what we're going to do to the Advanced Circuit Breaker.
* (@Brondahl) All the bits of tedious TODOs from the original PR comment.

**Comment by Brondahl** [comment]
> (@reisenberger + @Brondahl) Decide on what we're going to do to the Advanced Circuit Breaker.

So I see ... 3? ... options.

- Do nothing.
-- Assert that the Proportion-based approach of Adv.C.B. is incompatible with dynamic Half-Open durations, and leave it as it is.
-- :(
- Port identical functionality over.
-- Argue that the Advancedness of the Adv.C.B is solely about the details of when you intially call `Break_NeedsLock()` in the Controller, whilst in the `Closed` state, and the behaviour of the `HalfOpen` state is no different.
-- Simples!
- Create Advanced dynamic Breaks!?
-- Try to invent something that is analogous to the proportional healthchecks of the Closed state, and apply that to the Half-Open state.
-- Uhh ... sounds a bit rough.

My vote is (probably unsurprisingly) option No.2.
Opinions?

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (reisenberger) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: >comment was written the first time I tried to understand what was going on

No problem! Great diving already into how the circuit-breaker works!

New wording ☝️  sounds good to me! :+1:

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (reisenberger) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Yeah, that's exactly the dilemma: I went for/suggest `InvalidOperationException` as `ArgumentOutOfRangeException` inappropriate due to the out-of-range value not being a just-passed argument.

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (reisenberger) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Thanks @Brondahl for the factor out! I can only think of couple small things - can you push it? - easier to comment on pushed lines.  

Again, thanks for everything on this.

**Comment by reisenberger** [comment]
@Brondahl Over to you to push all the changes so far, I think I'm done on these:

(@reisenberger) **DONE** Sign-off on general review discussions.
(@reisenberger) **DONE** Confirm desired Exception type for `durationFactory() returns < 0" review markup.

(Might have small comments on `ExtensionUpToMaxValue()` once pushed.)

---

Re:

> (@reisenberger + @Brondahl) Decide on what we're going to do to the Advanced Circuit Breaker.

Great overview of options ^^ 👍 .  Reflecting on this.

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (reisenberger) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Again, thanks for the :+1: factoring out of `ExtensionUpToMaxValue(...)`.  Comments:

Let's avoid `System.Tuple`; we don't need to use a reference type here. We can stick to `System.ValueTuple`, just using the `(foo, bar)` value-tuple syntax.

Second, `DateTime.MinValue + TimeSpan.FromTicks(startPointTicks)` -> `new DateTime(startPointTicks)`.  (Given both the consumers of `ExtensionUpToMaxValue(...)` actually master their data (`_blockedTill`) in `Ticks`, it might be cleaner just to reduce the two overloads `ExtensionUpToMaxValue` to a single overload `ExtensionUpToMaxValue(long startPointTicks, TimeSpan proposedExtension`, but I'm not fussed if you want to push past this and leave the two overloads for now.)

**Comment by reisenberger** [comment]
>>(@reisenberger + @Brondahl) Decide on what we're going to do to the Advanced Circuit Breaker
>My vote is (probably unsurprisingly) option No.2.

Let us implementation option 2 at this time. Proportion-based health-checks for the half-open state (option 3) is a larger build, and requires wider consideration of what, conceptually, makes sense.

**Comment by Brondahl** [comment]
👍 Will update from the small comments above and push. Then start the rote changes. Will probably hold off writing completely new unit tests, and discuss them with you before I start to check that I'm approaching it correctly.

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Done.

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: >We can stick to System.ValueTuple
Ah, nice. Done.

>new DateTime(startPointTicks)
Ah, I looked for `DateTime.FromTicks()` but couldn't find it. Thank you :) Done.

I'm inclined to leave them separate. I think the rationale for the extension method is designed quite heavily around "we're representing DateTimes". If we switch to using Ticks everywhere, then we're *actually* using `long`s everywhere, and it will be tempting to start allowing anything up to `long.MaxValue` ... and then you'll end up with code to check that the `long` is a valid `DateTime` or `TimeSpan`.

Now that you've shown me the terse way to do Ticks => DateTime, I'd be happy to just have that inline in `PermitHalfOpenCircuitTest`, and drop the Ticks-based overload, so we only have the one, DateTime-based, method?

**Comment by Brondahl** [comment]
@reisenberger I've committed all the tweaks to the above, and the core change for the Adv. CB.
Shall I raise that as a new PR based off of this branch, or just push it as a change to this branch?
Which way to you prefer to view incremental changes on a PR?

**Comment by reisenberger** [comment]
>just push it as a change to this branch

... is fine. Thanks!

**Comment by Brondahl** [comment]
Heya, sorry for the delay. work got a little hectic as a project drew to a close.
Back on this now.

Have just pushed the aforementioned review markups and Implementation for Adv. CB.

Am working on the rote changes and similar.

**Comment by Brondahl** [comment]
@reisenberger I've completed the rote changes (adding the new options to all the syntaxes, and extending all the tests to run with static duration and also "dynamic" duration (i.e. where the `Func<int, TimeSpan>` syntax is used, but it always returns the same value.

Do you want to shout when you've reviewed the previous push ☝️ , and then I'll add these on top of it.

Alternatively you might just want to review my approach to the Test file updating.
I used a couple of regexes to Find'N'Replace all the changes

Make all the tests `Theories`, with `true` and `false` values:
 `[Fact]` => `[Theory, InlineData(true), InlineData(false)]` (not actually as a regex, just literals.)

Make all the tests parameterised to accept those values:
`( *public (void|async Task) \w*)\(\) *`
=>
`$1(bool timespanIsDynamic)`
_(and then manually fix the `Dispose()` method)_

Call the extensions and pass the parameter:
`\.(Advanced)?CircuitBreaker(Async)?\((.*)\);`
=>
`.$1CircuitBreaker$2_WithConditionalTimeSpan($3, timespanIsDynamic);`

\.(Advanced)?CircuitBreaker(Async)?\((.*)\);
.$1CircuitBreaker$2_WithConditionalTimeSpan($3, timespanIsDynamic);
_(plus some whitespace fixing for multi-line invocations in `AdvancedCBSpecs.cs`

**Comment by Brondahl** [comment]
Remaining actions:
* @reisenberger to re-review original markups (currently pushed)
* @reisenberger to shout if/when you want the rote changes pushed for review.
* @Brondahl to prepare [wiki-doc](https://github.com/App-vNext/Polly/wiki/Circuit-Breaker) updates.
* @Brondahl to figure out what specific testing is needed, and write up.
* @reisenberger to review test-plan.
* @Brondahl to implement testing, and progress from there.

**Comment by reisenberger** [comment]
@Brondahl Thanks for this!  I will review and get back to you (likely twds the end of the week or weekend)

**Comment by Brondahl** [comment]
Heya,

Had a proper look at the tests, and I don't think the tests for dynamic duration are too complex.
As a precursor, I'm going to factor out the most common, repetitive lines of the tests, after which the dynamic duration tests will look a bit like this:

```
        [Fact]
        public void Should_open_circuit_for_appropriate_dynamic_duration_when_call_raises_exception_repeatedly_in_sequence()
        {
            var startTime = 1.January(2000);
            var currentTime = startTime;
            SystemClock.UtcNow = () => currentTime;

            _breaker = Policy
                .Handle<DivideByZeroException>()
                .CircuitBreaker(2, (numberOfErrors) => TimeSpan.FromMinutes(numberOfErrors + 1));

            ExpectFailureCall_ToResultInPassThroughError____EndingUpInState(CircuitState.Closed);
            ExpectFailureCall_ToResultInPassThroughError____EndingUpInState(CircuitState.Open);

            ExpectFailureCall_ToResultInCircuitBreakerError_EndingUpInState(CircuitState.Open);

            // when initial duration has not passed, circuit still open
            currentTime = startTime.AddMinutes(1).AddMilliseconds(-1);
            ExpectState(CircuitState.Open);

            // when initial duration has passed, circuit now half-open
            currentTime = startTime.AddMinutes(1);
            ExpectState(CircuitState.HalfOpen);


            // Further error opens it back up, starting a new duration
            ExpectFailureCall_ToResultInPassThroughError____EndingUpInState(CircuitState.Open);
            var reopenTime = currentTime;


            // after initial duration has passed, circuit is *still* open, because the new duration is longer.
            currentTime = reopenTime.AddMinutes(2).AddMilliseconds(-1);
            ExpectState(CircuitState.Open);

            // once newer duration has passed, circuit is now half-open again.
            currentTime = reopenTime.AddMinutes(2);
            ExpectState(CircuitState.HalfOpen);
        }
```

Any concerns?
Any particular aspects of the functionality that you want me to ensure are being tested, which you think I might forget?

**Comment by reisenberger** [comment]
Thanks @Brondahl ! I'll aim to join thinking on specific test plans for dynamic-circuit-break-period either Wed or Thurs.

**Comment by reisenberger** [comment]
Hi @Brondahl 

### Proving existing behaviour unaffected

It looks like most of the existing tests on circuit-break duration are in a section marked `Circuit-breaker open->half-open->open/closed tests`.  There is also a separate test in the callback-delegates section about the break-duration being passed to the callback delegate: `Should_call_onbreak_with_the_correct_timespan`.  

For these existing tests, we would want variants checking that existing circuit-break-duration behaviour is not affected. IE If the tests do not already cover it, we should extend existing tests/add new tests confirming that when using policy configuration overloads with _constant_ durationOfBreak, the duration of break indeed remains constant when the circuit repeatedly transitions back to open from half-open.

### New tests

For any of the existing tests ^ which focused on break-duration, I guess we would want counterparts for the dynamic case.

**Comment by reisenberger** [comment]
#### Tests on orthogonal matters

If I've understood this correctly:

>Make all the tests Theories, with true and false values:
>[Fact] => [Theory, InlineData(true), InlineData(false)] (not actually as a regex, just literals.)

I don't think we need to do this.  Where a circuit-breaker tests is testing something entirely orthogonal to the changes you have made (that it breaks on the right exceptions; whatever), there's no need to duplicate these for both static and dynamic break durations.  If we'd taken that approach/took that approach going forward for each new feature added to circuit-breaker, we'd soon end up with a 2 ^ N explosion of tests  where N is the number of new features.

**Comment by Brondahl** [comment]
>I don't think we need to do this. Where a circuit-breaker tests is testing something entirely orthogonal to the changes you have made (that it breaks on the right exceptions; whatever), there's no need to duplicate these for both static and dynamic break durations.

I can see that argument.

I took my cue from the fact that each different variant on the CircuitBreaker Syntax, _does_ have a largely duplicated collection of tests (i.e. `CircuitBreakerSpecs`, `CircuitBreakTResultSpecs`, `CircuitBreakerAsyncSpecs`.)
Most of the tests in those files aren't testing things that are unique to the differences from the base implementation; they're just testing "this feature also works in the context of there being a return value in the policy", etc.

Similarly, all the tests are re-run for all the .NET versions.

_(Plus, I assumed you were in favour since you didn't comment on it in the initial review ;)  )_



I agree that the majority of that duplication is highly unlikely to excercise any new edge-cases.

For the avoidance of doubt, given that that change has been coded up for all Test Files and all tests pass, do you want me to remove those changes?
Happy to do so, if you'd still like me to, given the above :)

**Comment by Brondahl** [comment]
> It looks like most of the existing tests on circuit-break duration are in a section marked Circuit-breaker open->half-open->open/closed tests.
> [snip]
> For these existing tests, we would want variants checking that existing circuit-break-duration behaviour is not affected. IE If the tests do not already cover it, we should extend existing tests/add new tests confirming that when using policy configuration overloads with constant durationOfBreak, the duration of break indeed remains constant when the circuit repeatedly transitions back to open from half-open.

>For any of the existing tests ^ which focused on break-duration, I guess we would want counterparts for the dynamic case.

Yup, agree with all that.

**Comment by Brondahl** [comment]
>There is also a separate test in the callback-delegates section about the break-duration being passed to the callback delegate: Should_call_onbreak_with_the_correct_timespan.

Ah hah! Yes, that's the sort of thing I think I would have missed :)
Good catch; thanks!

**Comment by Brondahl** [comment]
LMK about removing the duplication of existing tests, and then I'll implement the extensions and new tests.

**Comment by reisenberger** [comment]
Thanks @Brondahl ! Yes, I think we should remove the dual-condition running of existing tests modelled with `[Theory, InlineData(true), InlineData(false)]`.  But thanks for suggesting it; I understand the motivation :+1:

File: src/Polly/CircuitBreaker/CircuitStateController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (reisenberger) [reply src/Polly/CircuitBreaker/CircuitStateController.cs]: Good points @Brondahl , nice changes :+1:

**Comment by reisenberger** [comment]
Thanks @Brondahl for everything on this :+1: . Re:

>@reisenberger to re-review original markups (currently pushed)

DONE - all good.

>@reisenberger to shout if/when you want the rote changes pushed for review.

READY

>@reisenberger to review test-plan.

DONE - as above

**Comment by Brondahl** [comment]
Rote changes to other syntaxes pushed.
As previously the first commit (`1974636a`) is duplicating the existing methods verbatim, so that the second commit (`fc80d1368`) has the interesting diffs. So you might want to exclude that commit from the diff you're looking at.

**Comment by Brondahl** [comment]
Updates to tests are not currently included since I need to revert the test duplication, as above. Once that's complete I'll push those changes, but since they're independant files, I don't think that should block review of the Syntax files.

**Comment by Brondahl** [comment]
Actions:

- @reisenberger to review rote syntax changes.
- @Brondahl to revert dynamic/static test duplication.
- @Brondahl to implement test refactoring across remaining test files.
- @Brondahl to implement new tests as per test plan above.
- @Brondahl to push all above test changes
- @reisenberger to review test changes.
- @Brondahl to prepare wiki-doc updates.

File: src/Polly/CircuitBreaker/AsyncCircuitBreakerTResultSyntax.cs [hunk 7]
```diff
@@ -18,55 +18,57 @@ private static void ValidateStaticDurationIsNonNegative(TimeSpan duration)
...
+        /// <param name="failureThreshold">The number of failures that are allowed before opening the circuit.</param>
         /// <param name="durationOfBreak">The duration the circuit will stay open before resetting.</param>
         /// <returns>The policy instance.</returns>
         /// <remarks>(see "Release It!" by Michael T. Nygard fi)</remarks>
-        /// <exception cref="ArgumentOutOfRangeException">exceptionsAllowedBeforeBreaking;Value must be greater than zero.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">failureThreshold;Value must be greater than zero.</exception>
         /// <exception cref="ArgumentOutOfRangeException">durationOfBreak;Value must be greater than zero</exception>
-        public static AsyncCircuitBreakerPolicy<TResult> CircuitBreakerAsync<TResult>(this PolicyBuilder<TResult> policyBuilder, int exceptionsAllowedBeforeBreaking, TimeSpan durationOfBreak)
+        public static AsyncCircuitBreakerPolicy<TResult> CircuitBreakerAsync<TResult>(this PolicyBuilder<TResult> policyBuilder, int failureThreshold, TimeSpan durationOfBreak)
```
Comment (reisenberger) [inline src/Polly/CircuitBreaker/AsyncCircuitBreakerTResultSyntax.cs]: I do like these parameter name changes, but technically, they are breaking the public API (users may have quoted the existing parameter names in the calls).  That would mean we would push the release of this feature into the forthcoming v8.0.0 rather than a v7.3.0.

If speed of release is important to you for the feature, you might want to revert the parameter name changes to avoid the breaking change.

File: src/Polly/CircuitBreaker/AdvancedCircuitBreakerSyntax.cs [hunk 8]
```diff
@@ -36,48 +42,41 @@ public static class AdvancedCircuitBreakerSyntax
...
+        /// again for a newly calculated duration; if no exception is thrown, the circuit will reset.
         /// </para>
         /// </summary>
         /// <param name="policyBuilder">The policy builder.</param>
         /// <param name="failureThreshold">The failure threshold at which the circuit will break (a number between 0 and 1; eg 0.5 represents breaking if 50% or more of actions result in a handled failure).</param>
         /// <param name="samplingDuration">The duration of the timeslice over which failure ratios are assessed.</param>
         /// <param name="minimumThroughput">The minimum throughput: this many actions or more must pass through the circuit in the timeslice, for statistics to be considered significant and the circuit-breaker to come into action.</param>
-        /// <param name="durationOfBreak">The duration the circuit will stay open before resetting.</param>
+        /// <param name="factoryForNextBreakDuration">A function to calculate the duration the circuit will stay open before resetting, based on the number of consecutive CircuitFailures. Re-evaluated each time the circuit Opens.</param>
```
[the remark this answers is not in the document]
Comment (Brondahl) [reply src/Polly/CircuitBreaker/AdvancedCircuitBreakerSyntax.cs]: 🤦‍♂ . Duh!
Yes, will adjust.

File: src/Polly/CircuitBreaker/AsyncCircuitBreakerTResultSyntax.cs (hunk 7 above)
Comment (Brondahl) [reply src/Polly/CircuitBreaker/AsyncCircuitBreakerTResultSyntax.cs]: Ah, shucks :)

Yup, I'll go and revert those and then have this change in a separate PR to go on afterwards.

**Comment by reisenberger** [comment]
>@reisenberger to review rote syntax changes.

DONE :+1: All good save minor parameter-naming things prev ^ mentioned. Thanks @Brondahl for all the great attention to detail on this PR!

**Comment by Brondahl** [comment]
Heya, Sorry about the radio silence; things got busy at work and in life in general :)

I've been gradually applying the general refactoring to the tests, to commonise the repetitive bits.
That's pretty slow going though. :(

So I'm going to put that down, and just apply the testing for ☝️ without that commonisation.
I'll raise a separate PR for the test commonisation (Like the param renaming), since I think that's worth cleaning up, still.

Hopefully I'll be able to get the Tests up by the end of the weekend.

**Comment by Brondahl** [comment]
Actions:

-     @Brondahl to implement new tests as per test plan above.
-     @Brondahl to push all above test changes
-     @reisenberger to review test changes.
-     @Brondahl to prepare wiki-doc updates.


=-=-=-=-=-=-=-=-=-=-=-=-=-=
> @Brondahl to revert dynamic/static test duplication.

Done.


> @Brondahl to implement test refactoring across remaining test files.

Deferred

**Comment by juliuscaesarnl** [comment]
Hi guys,

Great to see it is being implemented! I've been eagerly awaiting this feature.
I'm wondering what the ETA is, roughly, and whether or not it will be added in 7.3.0?

Thanks!

**Comment by Brondahl** [comment]
So, this got bogged down when I got to testing it.

The tests on Polly are comprehensive, but extremely verbose.
IIRC I started trying to refactor them, but got bogged down in the process, and ended up dropping this :(

I **believe** that the code itself is ready-to-go, so if someone wanted to pick up writing the verbose tests (or better doing the refactoring!) then it could be merged?

I dropped it partly because this was around a time when I got enthusiastic about 3 different OS changes, partly because I rolled off my project at the time and thus was no longer using Polly, and partly because my spare time got consumed by other non-programming stuff)

Realistically, I'm unlikely to come back to this anytime soon, I'm afraid. :(

**Comment by martincostello** [comment]
Closing due to inactivity.

Please open a new pull request if you would like to continue with a proposal for this change which we could consider for a v8 of Polly.

**Comment by baltermia** [comment]
This seems sooo close to being merged. Any updates??

**Comment by Brondahl** [comment]
@baltermia See my post from June 2021. That's still the status and is unlikely to change.
